File: harness/pi-runtime/features/codemode/src/kernels/jl/prelude.jl
# allow: SIZE_OK — this dependency-free kernel prelude is loaded as one runtime asset.
const SENPI_B64_ALPHABET = collect("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/")

function senpi_base64(text::AbstractString)
    bytes = codeunits(text)
    out = IOBuffer()
    index = 1
    while index <= length(bytes)
        first_byte = bytes[index]
        second_byte = index + 1 <= length(bytes) ? bytes[index + 1] : UInt8(0)
        third_byte = index + 2 <= length(bytes) ? bytes[index + 2] : UInt8(0)
        Base.write(out, SENPI_B64_ALPHABET[(first_byte >> 2) + 1])
        Base.write(out, SENPI_B64_ALPHABET[(((first_byte & 0x03) << 4) | (second_byte >> 4)) + 1])
        Base.write(out, index + 1 <= length(bytes) ? SENPI_B64_ALPHABET[(((second_byte & 0x0f) << 2) | (third_byte >> 6)) + 1] : '=')
        Base.write(out, index + 2 <= length(bytes) ? SENPI_B64_ALPHABET[(third_byte & 0x3f) + 1] : '=')
        index += 3
    end
    String(take!(out))
end

function senpi_status_events_enabled()
    get(senpi_connection, "statusEvents", true) !== false
end

function senpi_emit_status(op::AbstractString, fields::AbstractDict=Dict{String, Any}(); force::Bool=false)
    (force || senpi_status_events_enabled()) || return nothing
    event = Dict{String, Any}("op" => string(op))
    for (key, value) in fields
        event[string(key)] = value
    end
    senpi_emit(Dict("type" => "status", "event" => event))
    nothing
end

function senpi_with_bridge_timeout_pause(operation::Function)
    senpi_emit_status("timeout-pause"; force=true)
    try
        return operation()
    finally
        senpi_emit_status("timeout-resume"; force=true)
    end
end

function senpi_url_decode(value::AbstractString)
    out = IOBuffer()
    index = 1
    while index <= ncodeunits(value)
        character = Char(codeunit(value, index))
        if character == '%' && index + 2 <= ncodeunits(value)
            hex = value[index + 1:index + 2]
            parsed = tryparse(UInt8, hex; base=16)
            if parsed !== nothing
                Base.write(out, parsed)
                index += 3
                continue
            end
        end
        Base.write(out, character)
        index += 1
    end
    String(take!(out))
end

function senpi_resolve_path(value::AbstractString)
    raw = string(value)
    matched = match(r"^([a-z][a-z0-9+.\-]*)://(.*)$"i, raw)
    matched === nothing && return abspath(raw)
    scheme = lowercase(string(matched.captures[1]))
    roots = get(senpi_connection, "localRoots", nothing)
    root = roots isa AbstractDict ? get(roots, scheme, nothing) : nothing
    root isa AbstractString && !isempty(root) || error("Protocol paths are not supported by this helper: $raw")
    relative = senpi_url_decode(replace(string(matched.captures[2]), '\\' => '/'))
    root_path = abspath(string(root))
    isempty(relative) && return root_path
    (startswith(relative, '/') || ".." in split(relative, '/')) && error("Unsafe $scheme:// path (absolute or traversal): $raw")
    resolved = abspath(joinpath(root_path, relative))
    (resolved == root_path || startswith(resolved, root_path * string(Base.Filesystem.path_separator))) || error("$scheme:// path escapes its root: $raw")
    resolved
end

function senpi_display_payload(value)
    if value isa AbstractDict
        kind = get(value, "type", get(value, :type, nothing))
        text_value = get(value, "text", get(value, :text, nothing))
        if kind == "markdown" && text_value !== nothing
            return "text/markdown", string(text_value), false
        end
        mime_type = get(value, "mimeType", get(value, :mimeType, nothing))
        data = get(value, "data", get(value, :data, nothing))
        if kind == "image" && mime_type isa AbstractString && data !== nothing
            return string(mime_type), string(data), true
        end
        return "application/json", senpi_json(value), false
    end
    value isa AbstractVector && return "application/json", senpi_json(value), false
    "text/plain", string(value), false
end

function display(value)
    mime_type, payload, encoded = senpi_display_payload(value)
    senpi_emit(Dict("type" => "display", "mimeType" => mime_type, "dataBase64" => encoded ? payload : senpi_base64(payload)))
    nothing
end

function display_image(base64_value::AbstractString, mime_type::AbstractString="image/png")
    senpi_emit(Dict("type" => "display", "mimeType" => string(mime_type), "dataBase64" => string(base64_value)))
    nothing
end

function text(value)
    senpi_emit(Dict("type" => "text", "stream" => "stdout", "data" => string(value)))
    nothing
end

function print(values...)
    text(join(string.(values), ""))
end

function read(path::AbstractString, offset::Integer=1, limit::Union{Integer, Nothing}=nothing)
    resolved = senpi_resolve_path(path)
    content = Base.read(resolved, String)
    if offset > 1 || limit !== nothing
        lines = split(content, '\n'; keepempty=true)
        start = max(1, Int(offset))
        finish = limit === nothing ? length(lines) : min(length(lines), start + Int(limit) - 1)
        content = start <= length(lines) ? join(lines[start:finish], '\n') : ""
    end
    senpi_emit_status("read", Dict("path" => resolved, "chars" => length(content), "preview" => first(content, min(length(content), 500))))
    content
end

function write(path::AbstractString, content)
    resolved = senpi_resolve_path(path)
    mkpath(dirname(resolved))
    data = string(content)
    open(resolved, "w") do io
        Base.write(io, data)
    end
    senpi_emit_status("write", Dict("path" => resolved, "chars" => length(data)))
    resolved
end

function env(key=nothing, value=nothing)
    if key === nothing
        values = Dict{String, String}(string(name) => string(item) for (name, item) in ENV)
        senpi_emit_status("env", Dict("count" => length(values), "keys" => sort(collect(keys(values)))[1:min(20, length(values))]))
        return values
    end
    name = string(key)
    if value === nothing
        resolved = get(ENV, name, nothing)
        senpi_emit_status("env", Dict("key" => name, "value" => resolved, "action" => "get"))
        return resolved
    end
    resolved = string(value)
    ENV[name] = resolved
    senpi_emit_status("env", Dict("key" => name, "value" => resolved, "action" => "set"))
    resolved
end

struct SenpiToolProxy end

struct SenpiToolCallable
    name::String
end

function (callable::SenpiToolCallable)(args=nothing; kwargs...)
    values = Dict{String, Any}()
    if args !== nothing
        args isa AbstractDict || error("tool.$(callable.name)(...) expects a dictionary of arguments")
        for (key, value) in args
            values[string(key)] = value
        end
    end
    for (key, value) in kwargs
        values[string(key)] = value
    end
    senpi_with_bridge_timeout_pause(() -> senpi_call_tool(callable.name, values))
end

Base.getproperty(::SenpiToolProxy, name::Symbol) = SenpiToolCallable(string(name))
const tool = SenpiToolProxy()

function completion(prompt::AbstractString; model="default", system=nothing, schema=nothing, kwargs...)
    options = Dict{String, Any}("model" => model)
    system !== nothing && (options["system"] = system)
    schema !== nothing && (options["schema"] = schema)
    for (key, value) in kwargs
        options[string(key)] = value
    end
    response = senpi_with_bridge_timeout_pause(() -> senpi_completion(string(prompt), options))
    response isa AbstractDict || return response
    haskey(response, "value") && return response["value"]
    get(response, "text", response)
end

function tool_schema(name=nothing)
    arguments = Dict{String, Any}()
    name !== nothing && (arguments["name"] = string(name))
    senpi_with_bridge_timeout_pause(() -> senpi_call_tool("__schema__", arguments))
end

function output(ids...; format="raw", offset=nothing, limit=nothing)
    isempty(ids) && error("At least one output ID is required")
    format in ("raw", "tail") || error("output() format must be 'raw' or 'tail'")
    arguments = Dict{String, Any}("ids" => string.(ids), "format" => format)
    offset !== nothing && (arguments["offset"] = offset)
    limit !== nothing && (arguments["limit"] = limit)
    senpi_with_bridge_timeout_pause(() -> senpi_call_tool("__output__", arguments))
end

function agent(prompt::AbstractString; agent="task", model=nothing, label=nothing, schema=nothing, isolated=nothing, apply=nothing, merge=nothing, handle=false, kwargs...)
    arguments = Dict{String, Any}("prompt" => string(prompt), "agent" => agent)
    for (key, value) in (("model", model), ("label", label), ("schema", schema), ("isolated", isolated), ("apply", apply), ("merge", merge))
        value !== nothing && (arguments[key] = value)
    end
    for (key, value) in kwargs
        arguments[string(key)] = value
    end
    handle && (arguments["handle"] = true)
    response = senpi_with_bridge_timeout_pause(() -> senpi_call_tool("__agent__", arguments))
    record = response isa AbstractDict ? response : Dict{String, Any}()
    text_value = get(record, "text", response)
    parsed = schema === nothing ? text_value : haskey(record, "data") ? record["data"] : senpi_json_parse(string(text_value))
    handle || return parsed
    result = Dict{String, Any}("text" => text_value, "output" => text_value, "id" => get(record, "id", nothing), "agent" => get(record, "agent", agent))
    result["handle"] = get(record, "handle", result["id"] === nothing ? nothing : "agent://" * string(result["id"]))
    schema !== nothing && (result["data"] = parsed)
    result
end

function senpi_pool_map(items, callback)
    values = collect(items)
    isempty(values) && return Any[]
    configured_width = get(senpi_connection, "parallelPoolWidth", 4)
    width = configured_width isa Real && isfinite(configured_width) ? Int(floor(configured_width)) : 4
    workers = min(max(width, 1), length(values))
    tokens = Channel{Nothing}(workers)
    for _ in 1:workers
        put!(tokens, nothing)
    end
    results = Vector{Any}(undef, length(values))
    errors = Dict{Int, Any}()
    lock = ReentrantLock()
    @sync for index in eachindex(values)
        @async begin
            take!(tokens)
            try
                results[index] = callback(values[index])
            catch error
                Base.lock(lock) do
                    errors[index] = error
                end
            finally
                put!(tokens, nothing)
            end
        end
    end
    isempty(errors) || throw(errors[minimum(keys(errors))])
    results
end

function parallel(thunks)
    values = collect(thunks)
    for thunk in values
        applicable(thunk) || error("parallel() expects zero-argument callables")
    end
    senpi_pool_map(values, thunk -> thunk())
end

function pipeline(items, stages...)
    values = collect(items)
    for stage in stages
        isempty(values) && return values
        applicable(stage, first(values)) || error("pipeline() stages must be callables")
        values = senpi_pool_map(values, stage)
    end
    values
end

function log(message)
    senpi_emit(Dict("type" => "log", "message" => string(message)))
    nothing
end

function phase(title)
    senpi_emit(Dict("type" => "phase", "title" => string(title)))
    nothing
end

File: harness/pi-runtime/features/codemode/src/kernels/rb/prelude.rb
require "base64"
require "fileutils"
require "json"
require "uri"

SENPI_RESERVED_AGENT_TOOL = "__agent__"
SENPI_RESERVED_OUTPUT_TOOL = "__output__"
SENPI_RESERVED_SCHEMA_TOOL = "__schema__"
SENPI_INTERNAL_URL = Regexp.new("\\A([a-z][a-z0-9+.\\-]*)://(.*)\\z", Regexp::IGNORECASE)

def __senpi_status_enabled?
  connection = $__senpi_connection
  return true unless connection.is_a?(Hash)
  connection["statusEvents"] != false
end

def __senpi_emit_status(op, fields = {}, force: false)
  return unless force || __senpi_status_enabled?
  __senpi_emit({ "type" => "status", "event" => { "op" => op }.merge(fields.transform_keys(&:to_s)) })
end

def __senpi_resolve_path(value)
  raw = value.to_s
  match = SENPI_INTERNAL_URL.match(raw)
  return File.expand_path(raw) unless match

  scheme = match[1].downcase
  roots = $__senpi_connection.is_a?(Hash) ? $__senpi_connection["localRoots"] : nil
  root = roots[scheme] if roots.is_a?(Hash)
  raise "Protocol paths are not supported by this helper: #{raw}" unless root.is_a?(String) && !root.empty?

  relative = URI::DEFAULT_PARSER.unescape(match[2].tr("\\", "/"))
  root_path = File.expand_path(root)
  return root_path if relative.empty?
  if relative.start_with?("/") || relative.split("/").include?("..")
    raise "Unsafe #{scheme}:// path (absolute or traversal): #{raw}"
  end

  resolved = File.expand_path(relative, root_path)
  unless resolved == root_path || resolved.start_with?(root_path + File::SEPARATOR)
    raise "#{scheme}:// path escapes its root: #{raw}"
  end
  resolved
end

def __senpi_display_payload(value)
  if value.is_a?(Hash)
    kind = value["type"] || value[:type]
    text_value = value["text"] || value[:text]
    return ["text/markdown", text_value.to_s] if kind == "markdown" && !text_value.nil?
    return ["image/png", value["data"].to_s] if kind == "image" && value["mimeType"] == "image/png"
    return ["image/jpeg", value["data"].to_s] if kind == "image" && value["mimeType"] == "image/jpeg"
    return ["application/json", JSON.generate(value)]
  end
  return ["application/json", JSON.generate(value)] if value.is_a?(Array)
  ["text/plain", value.to_s]
end

def display(value)
  mime_type, payload = __senpi_display_payload(value)
  data_base64 = mime_type.start_with?("image/") ? payload : Base64.strict_encode64(payload)
  __senpi_emit({ "type" => "display", "mimeType" => mime_type, "dataBase64" => data_base64 })
  nil
end

def display_image(base64, mime_type: "image/png")
  __senpi_emit({ "type" => "display", "mimeType" => mime_type.to_s, "dataBase64" => base64.to_s })
  nil
end

def text(data)
  __senpi_emit({ "type" => "text", "stream" => "stdout", "data" => data.to_s })
  nil
end

def print(*values)
  text(values.join)
end

def read(path, offset = 1, limit = nil)
  resolved = __senpi_resolve_path(path)
  data = File.read(resolved, encoding: Encoding::UTF_8)
  if offset > 1 || !limit.nil?
    lines = data.lines
    start = [offset.to_i - 1, 0].max
    data = lines[start, limit || lines.length].to_a.join
  end
  __senpi_emit_status("read", { "path" => resolved, "chars" => data.length, "preview" => data[0, 500].to_s })
  data
end

def write(path, content)
  resolved = __senpi_resolve_path(path)
  FileUtils.mkdir_p(File.dirname(resolved))
  data = content.to_s
  File.write(resolved, data)
  __senpi_emit_status("write", { "path" => resolved, "chars" => data.length })
  resolved
end

def env(key = nil, value = nil)
  if key.nil?
    entries = ENV.to_h.sort.to_h
    __senpi_emit_status("env", { "count" => entries.length, "keys" => entries.keys.first(20) })
    return entries
  end

  name = key.to_s
  if value.nil?
    resolved = ENV[name]
    __senpi_emit_status("env", { "key" => name, "value" => resolved, "action" => "get" })
    return resolved
  end

  resolved = value.to_s
  ENV[name] = resolved
  __senpi_emit_status("env", { "key" => name, "value" => resolved, "action" => "set" })
  resolved
end

def __senpi_bridge_request(path, payload)
  connection = $__senpi_connection
  raise "Ruby tool bridge is not initialized" unless connection.is_a?(Hash)
  port = connection["port"]
  token = connection["token"]
  raise "Ruby tool bridge is not initialized" unless port.is_a?(Integer) && token.is_a?(String)

  uri = URI("http://127.0.0.1:#{port}#{path}")
  request = Net::HTTP::Post.new(uri)
  request["authorization"] = "Bearer #{token}"
  request["content-type"] = "application/json"
  request.body = JSON.generate(payload)
  __senpi_emit_status("timeout-pause", force: true)
  begin
    response = Net::HTTP.start(uri.hostname, uri.port, open_timeout: 10, read_timeout: 60) { |http| http.request(request) }
  ensure
    __senpi_emit_status("timeout-resume", force: true)
  end
  body = JSON.parse(response.body.to_s)
  return body["value"] if body.is_a?(Hash) && body["ok"] == true

  error = body.is_a?(Hash) ? body["error"] : body
  raise(error.is_a?(Hash) ? error["message"].to_s : error.to_s)
end

def __senpi_call_tool(name, args)
  __senpi_bridge_request("/call", { "callId" => "rb-#{Process.pid}-#{rand(1_000_000)}", "toolName" => name, "args" => args })
end

class SenpiToolCallable
  def initialize(name)
    @name = name
  end

  def call(args = nil, **kwargs)
    merged = args.nil? ? {} : args.is_a?(Hash) ? args.transform_keys(&:to_s) : raise(ArgumentError, "tool.#{@name}(...) expects a Hash of arguments")
    kwargs.each { |key, value| merged[key.to_s] = value }
    __senpi_call_tool(@name, merged)
  end
end

class SenpiToolProxy < BasicObject
  def method_missing(name, args = nil, **kwargs)
    ::SenpiToolCallable.new(name.to_s).call(args, **kwargs)
  end

  def [](name)
    ::SenpiToolCallable.new(name.to_s)
  end

  def respond_to_missing?(_name, _private = false)
    true
  end
end

def tool
  $__senpi_tool_proxy ||= SenpiToolProxy.new
end

def completion(prompt, model: "default", system: nil, schema: nil, **kwargs)
  options = { "model" => model }.merge(kwargs.transform_keys(&:to_s))
  options["system"] = system unless system.nil?
  options["schema"] = schema unless schema.nil?
  result = __senpi_bridge_request("/completion", { "prompt" => prompt.to_s, "opts" => options })
  return result unless result.is_a?(Hash)
  return result["value"] if result.key?("value")

  result.fetch("text", result)
end

def tool_schema(name = nil)
  args = name.nil? ? {} : { "name" => name.to_s }
  __senpi_call_tool(SENPI_RESERVED_SCHEMA_TOOL, args)
end

def output(*ids, format: "raw", offset: nil, limit: nil)
  raise ArgumentError, "At least one output ID is required" if ids.empty?
  raise ArgumentError, "output() format must be 'raw' or 'tail'" unless ["raw", "tail"].include?(format)
  args = { "ids" => ids.map(&:to_s), "format" => format }
  args["offset"] = offset unless offset.nil?
  args["limit"] = limit unless limit.nil?
  __senpi_call_tool(SENPI_RESERVED_OUTPUT_TOOL, args)
end

def agent(prompt, agent: "task", model: nil, label: nil, schema: nil, isolated: nil, apply: nil, merge: nil, handle: false)
  args = { "prompt" => prompt.to_s, "agent" => agent }
  { "model" => model, "label" => label, "schema" => schema, "isolated" => isolated, "apply" => apply, "merge" => merge }.each do |key, value|
    args[key] = value unless value.nil?
  end
  args["handle"] = true if handle
  response = __senpi_call_tool(SENPI_RESERVED_AGENT_TOOL, args)
  record = response.is_a?(Hash) ? response : {}
  text_value = record.fetch("text", response)
  result = schema.nil? ? text_value : record.key?("data") ? record["data"] : JSON.parse(text_value.to_s)
  return result unless handle
  { "text" => text_value, "output" => text_value, "handle" => record["handle"] || (record["id"] && "agent://#{record["id"]}"), "id" => record["id"], "agent" => record.fetch("agent", agent) }.tap do |node|
    node["data"] = result unless schema.nil?
  end
end

def __senpi_pool_map(items)
  values = items.to_a
  return [] if values.empty?
  connection = $__senpi_connection
  configured_width = connection.is_a?(Hash) ? connection["parallelPoolWidth"] : nil
  width = configured_width.is_a?(Numeric) ? configured_width.to_i : 4
  workers = [[width, 1].max, values.length].min
  results = Array.new(values.length)
  failures = {}
  failure_mutex = Mutex.new
  queue = Queue.new
  values.each_index { |index| queue << index }
  threads = workers.times.map do
    Thread.new do
      loop do
        index = queue.pop(true) rescue nil
        break if index.nil?
        begin
          results[index] = yield(values[index])
        rescue Exception => error
          failure_mutex.synchronize { failures[index] = error }
        end
      end
    end
  end
  threads.each(&:join)
  raise failures[failures.keys.min] unless failures.empty?
  results
end

def parallel(thunks)
  values = thunks.to_a
  values.each do |thunk|
    raise TypeError, "parallel() expects an iterable of zero-arg callables" unless thunk.respond_to?(:call)
  end
  __senpi_pool_map(values) { |thunk| thunk.call }
end

def pipeline(items, *stages)
  values = items.to_a
  stages.each do |stage|
    raise TypeError, "pipeline() stages must be callables" unless stage.respond_to?(:call)
    values = __senpi_pool_map(values) { |value| stage.call(value) }
  end
  values
end

def log(message)
  __senpi_emit({ "type" => "log", "message" => message.to_s })
  nil
end

def phase(title)
  __senpi_emit({ "type" => "phase", "title" => title.to_s })
  nil
end
